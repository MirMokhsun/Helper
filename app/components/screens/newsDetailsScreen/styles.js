import { StyleSheet } from 'react-native';
import Utils from '../../../api/utils/utils';
const { width } = Utils.size;

export const styles = StyleSheet.create({
    stillNewscontainer: {
        flex: 1,
        width: '100%',
    },
    stillNewsTitleText: {
        fontFamily: 'Roboto-Light',
        fontSize: 16,
        color: '#1459D2',
        lineHeight: 40,
    },
    stillNewsTitleWrapper: {
        height: 40,
    },
    wrapperSelectNews: {
        flex: 1,
        marginTop: 10,
        width: width - 20,
        backgroundColor: '#FFFFFF'
    },
    wrapperStillNews: {
        width: width - 20,
        backgroundColor: '#ffffff'
    },
    wrapper: {
        width: width,
        // height: height-40,
        height: '100%',
        alignItems: "center",
        backgroundColor: '#ffffff'
    },
    imageSelectStyle: {
        width: '100%',
        height: 200,
    },
    imageContainner: {
        height: 200,
    }, textNewsStyle: {
        marginTop: 15,
        marginBottom: 10,
        fontFamily: 'Roboto-Light',
        fontSize: 16,
        lineHeight: 20,
        color: '#000756'
    }, timeNewsStyle: {
        fontFamily: 'Roboto-Light',
        fontSize: 11,
        lineHeight: 13,
        color: '#1459D2'
    }, marginText: {
        marginLeft: 15,
    }, timeNewsWrapper: {
        marginTop: 10,
        marginLeft: 5,
        flexDirection: 'row',
    },
    maincontainer: {
        width: '100%',
    }, imageStyle: {
        position: "absolute",
        top: -70,
        width: '100%',
        height: 300,
    },
});