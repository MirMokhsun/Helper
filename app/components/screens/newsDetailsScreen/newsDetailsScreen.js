/* eslint-disable no-plusplus */
import React from 'react';
import { View, Platform, Image, ScrollView, Text, I18nManager } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import SmallClockGray from '../tabNavigatorScreens/newsScreen/smallNew/svg/smallClockGray';
import ApiForNews from '../../../api/processingNews/processingForNews';
import SmallNew from '../tabNavigatorScreens/newsScreen/smallNew/smallNew';
import Utils from '../../../api/utils/utils';
import { styles } from './styles';
import { localization } from '../../../common/localization';
import { arrayClosedDefaultImages } from '../tabNavigatorScreens/newsScreen/images/defaultClosedImages/arrayClosedDefaultImages';
import { arrayOpenDefaultOpenImages } from '../tabNavigatorScreens/newsScreen/images/defaultOpenImages/arrayOpenDefaultImages';

const { width } = Utils.size;
const { setTextAlign } = Utils;

export class NewsDetailsScreen extends React.Component {
    static navigationOptions = ({ navigation }) => {
        return {
            headerTintColor: '#000756',
            headerBackTitle: null,
            title: navigation.state.params ? navigation.state.params.title ? navigation.state.params.title : 'News' : 'News',
            headerTitleStyle: {
                color: '#000756',
                alignSelf: 'center',
                width: '100%',
                height: '100%',
                textAlign: "center",
                fontSize: 20,
                lineHeight: 40,
                ...Platform.select({
                    ios: {
                        paddingRight: 0,
                    },
                    android: {
                        paddingRight: I18nManager.isRTL ? 0 : 30,
                        paddingLeft: I18nManager.isRTL ? 30 : 0,
                    },
                }),
                fontFamily: 'Roboto',
                backgroundColor: "#FFFFFF",
            },
            headerStyle: {
                backgroundColor: "#FFFFFF",
                borderBottomColor: "#FFFFFF",
                ...Platform.select({
                    ios: {
                    },
                    android: {
                        marginTop: 0,
                        paddingTop: 10,
                    },
                }),
            },
            headerRight: (<View />),
        }
    };

    constructor(props) {
        super(props)
        this.newsObj = new ApiForNews();
        const { language } = this.props;
        this.state = {
            language,
            selectRenderNews: null,
            stillNews: null,
            currentTextAlign: setTextAlign(language)
        }
        this.maxCount = 5;
        this.countRandom = 20;
    }

    componentDidUpdate(prevProps) {
        const { language } = this.props;
        const { language: prevlanguage } = prevProps;

        if (language !== prevlanguage) {
            this.setState({ currentTextAlign: setTextAlign(language) });
        }
    }

    componentWillMount() {
        const { navigation, language } = this.props;
        navigation.setParams({ title: localization[language]["News"] });
    }

    selectionNews = (news, id) => {
        const result = {};
        for (const key in news) {
            if (news[key].ID === id) {
                result.selectNews = news[key];
                result.key = key;
                break;
            }
        }
        return result;
    }

    renderOpenNews = (selectNews, randomNumb) => {
        let result = null;
        let defaultImages = null;
        const { currentTextAlign, language } = this.state;
        if (selectNews) {
            const date = this.newsObj.selectLocationData( language, selectNews["PublishDate"]);
            const urlImajes = selectNews["Image"];
            const isDefoultImages = this.newsObj.isDefoultImages(urlImajes);
            if (!isDefoultImages) {
                defaultImages = arrayOpenDefaultOpenImages[randomNumb];
            }
            result = (
                <View style={styles.maincontainer}>
                    <View style={styles.imageContainner}>
                        {isDefoultImages ? <Image source={{ uri: urlImajes }} resizeMode="stretch" style={styles.imageSelectStyle} /> : defaultImages}
                    </View>
                    <View style={styles.timeNewsWrapper}>
                        <SmallClockGray />
                        <Text style={styles.timeNewsStyle}>{date}</Text>
                    </View>
                    <Text style={[styles.textNewsStyle, { textAlign: currentTextAlign }]}>{selectNews["Encoded"]}</Text>
                </View>)
        }
        return result;
    }

    renderStillNews = (news, counterIndex, maxCount, newsArrayLength) => {
        let stillNews = null; let defaultImages = null; let randomNumb = 1;
        const { currentTextAlign, language } = this.state;
        if (news && counterIndex && maxCount && newsArrayLength) {
            stillNews = [];
            while (stillNews.length !== maxCount) {
                if (counterIndex === newsArrayLength) {
                    counterIndex = 0;
                }
                const date = this.newsObj.selectLocationData( language, news[counterIndex]["PublishDate"]);
                const urlImajes = news[counterIndex]["Image"];
                const isDefoultImages = this.newsObj.isDefoultImages(urlImajes);
                const cutTextNews = this.newsObj.getTwentyWords(news[counterIndex]["Description"], 20);
                if (!isDefoultImages) {
                    randomNumb = this.newsObj.getRandomNumb(this.countRandom);
                    defaultImages = arrayClosedDefaultImages[randomNumb];
                }
                stillNews.push(
                    <SmallNew
                        scrollView={this.refs.scrollView}
                        randomNumb={randomNumb}
                        id={news[counterIndex]["ID"]}
                        key={counterIndex}
                        image={isDefoultImages ? <Image source={{ uri: urlImajes }} resizeMode="stretch" style={styles.imageStyle} /> : defaultImages}
                        textNews={cutTextNews}
                        timeNews={date}
                        currentTextAlign={currentTextAlign}
                    />
                );
                counterIndex+=1;
            }
        }
        return stillNews;
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        const { news, id } = nextProps;
        const { randomNumb } = nextProps.navigation.state.params;
        const selectionNewsObj = this.selectionNews(news, id);
        const selectRenderNews = this.renderOpenNews(selectionNewsObj.selectNews, randomNumb);
        const stillNews = this.renderStillNews(news, ++selectionNewsObj.key, this.maxCount, news.length);
        this.setState({
            selectRenderNews,
            stillNews
        });
    }

    render() {
        const { stillNews, selectRenderNews, currentTextAlign } = this.state;
        const { language } = this.props;
        return (
            <View style={styles.wrapper}>
                <ScrollView ref="scrollView" contentContainerStyle={{ width }}>
                    <View style={styles.wrapperSelectNews} alignSelf="center">
                        {selectRenderNews}
                    </View>
                    <View style={styles.wrapperStillNews} alignSelf="center">
                        <View style={styles.stillNewsTitleWrapper}>
                            <Text style={[styles.stillNewsTitleText, { textAlign: currentTextAlign }]}>
                                {localization[language]['More news']}
                            </Text>
                        </View>
                        <View style={styles.stillNewscontainer}>
                            {stillNews}
                        </View>
                    </View>
                </ScrollView>
            </View>)
    }
}

const mapStateToProps = state => ({
    news: state.setNews.news,
    language: state.currentLanguage.language,
    id: state.setNewsDetails.id
});
const mapDispatchToProps = dispatch => bindActionCreators({
}, dispatch);
export default connect(mapStateToProps, mapDispatchToProps)(NewsDetailsScreen);