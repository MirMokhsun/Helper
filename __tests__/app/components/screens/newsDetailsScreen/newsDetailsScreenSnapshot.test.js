import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { NewsDetailsScreen } from '../../../../../app/components/screens/newsDetailsScreen/newsDetailsScreen';

const mockNews = [{
    Channels: [{ ID: 36, Name: "Umarkets" }],
    Comments: "https://news.umarkets.org/feed/En/surplus-budget-uk-reached-in-january-historical-record",
    CommentsLanguage: ["https://news.umarkets.org/feed/Ru/surplus-budget-uk-reached-in-january-historical-record", "https://news.umarkets.org/feed/Ar/surplus-budget-uk-reached-in-january-historical-record", "https://news.umarkets.org/feed/Pl/surplus-budget-uk-reached-in-january-historical-record"],
    Creator: "EvgeniaT",
    Description: "The UK budget surplus in January reached a record size and, according to the British Office for National Statistics, amounted to 14 billion 900 million pounds. Revenues to the budget increased in annual terms by 9.8%. This was facilitated by an increase in income tax collections amid record-breaking low ...",
    Encoded: "The UK budget surplus in January reached a record size and, according to the British Office for National Statistics, amounted to 14 billion 900 million pounds. Revenues to the budget increased in annual terms by 9.8%. This was facilitated by an increase in income tax collections amid record-breaking low unemployment. In addition, VAT revenues increased due to higher consumer spending. In January, there was a decrease in budget expenditures by 3% and a fall in public debt by 17%. Following the results of 10 months of the fiscal year, which ends in March, the British budget deficit reached the minimum level of 18 billion 200 million pounds in 18 years, having decreased by 47% over the year. At the end of January, the national debt of the country has decreased to the lowest since May 2012, to 74% of GDP.",
    ID: 47322,
    Image: "https://media-news.umarkets.org/admin/img/01929726-bf22-4e7f-ae5d-cb531ca2408d",
    Lang: { Code: "En", ID: 2 },
    Name: "English",
    Link: "https://www.umarkets.org/en/news/surplus-budget-uk-reached-in-january-historical-record",
    PublishDate: 1550820994,
    Tags: [{ ID: 11, Name: "Forex" }, { ID: 10, Name: "Economic" }],
    Title: "UK budget surplus hit record high in January"
}];

it('create snapshot NewsDetailsScreen', () => {
    const renderer = new ShallowRenderer()
    let toolbarSnapshot = renderer.render(<NewsDetailsScreen
        id={47322}
        language={"ru"}
        navigation={{ navigate: () => { }, setParams: () => { }, state: { params: 'test' } }}
        news={mockNews}
    />);
    expect(toolbarSnapshot).toMatchSnapshot();
});
