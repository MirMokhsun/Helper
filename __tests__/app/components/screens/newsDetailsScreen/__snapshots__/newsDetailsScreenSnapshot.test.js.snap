// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create snapshot NewsDetailsScreen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "height": "100%",
      "width": 750,
    }
  }
>
  <ScrollViewMock
    contentContainerStyle={
      Object {
        "width": 750,
      }
    }
  >
    <View
      alignSelf="center"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "marginTop": 10,
          "width": 730,
        }
      }
    />
    <View
      alignSelf="center"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "width": 730,
        }
      }
    >
      <View
        style={
          Object {
            "height": 40,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1459D2",
                "fontFamily": "Roboto-Light",
                "fontSize": 16,
                "lineHeight": 40,
              },
              Object {
                "textAlign": "left",
              },
            ]
          }
        >
          Ещё новости
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "width": "100%",
          }
        }
      />
    </View>
  </ScrollViewMock>
</View>
`;
